include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+						auto stageString = std::to_string(stage);
+						hashable.insert(hashable.end(), stageString.data()[0], stageString.data()[stageString.size()]);
+						auto versionString = std::to_string(static_cast<uint32_t>(targetSpirvVersion));
+						hashable.insert(hashable.end(), versionString.data()[0], versionString.data()[versionString.size()]);
+						auto debugString = std::to_string(static_cast<uint8_t>(debugInfoFlags.value));
+						hashable.insert(hashable.end(), debugString.data()[0], debugString.data()[debugString.size()]);
+						for (auto pass : optimizerPasses) {
+							auto passString = std::to_string(pass);
+							hashable.insert(hashable.end(), passString.data()[0], passString.data()[passString.size()]);
+						}
+						return hashable;
+					}
+
+					IShader::E_SHADER_STAGE stage;
+					E_SPIRV_VERSION targetSpirvVersion;
+					std::vector<ISPIRVOptimizer::E_OPTIMIZER_PASS> optimizerPasses;
+					core::bitflag<E_DEBUG_INFO_FLAGS> debugInfoFlags;
+					SPreprocessorData preprocessorData;
+				};
+
+				// The ordering is important here, the dependencies MUST be added to the array IN THE ORDER THE PREPROCESSOR INCLUDED THEM!
+				using dependency_container_t = std::vector<SPreprocessingDependency>;
+				inline SEntry(std::string_view _mainFileContents, dependency_container_t&& _dependencies, SCompilerData&& compilerData) : mainFileContents(std::move(std::string(_mainFileContents))), dependencies(std::move(_dependencies))
+				{
+				}
+
+				// Needed to get the vector deserialization automatically
+				SEntry() {}
+				inline SEntry(SEntry&&) = default;
+				inline SEntry& operator=(SEntry&&) = default;
+
+				// This next bit is a bit of a Frankenstein. We serialize shader creation parameters into a json, while the actual shader code goes in another file
+				struct CPUShaderCreationParams {
+					IShader::E_SHADER_STAGE stage;
+					IShader::E_CONTENT_TYPE contentType; //I think this one could be skipped since it's always going to be SPIR-V
+					std::string filepathHint;
+					uint64_t codeByteSize = 0;
+					uint64_t offset = 0; // Offset into the serialized .bin for the Cache where code starts
+				};
+
+				std::string mainFileContents;
+				SCompilerData compilerData;
+				// Keeping this one commented out. Could be useful for lazy-loading of dependencies if it takes up too much memory
+				// uint64_t entryID;
+
+				dependency_container_t dependencies;
+				mutable CPUShaderCreationParams shaderParams;
+				bool serialized = false;
+				mutable core::smart_refctd_ptr<asset::ICPUShader> value = nullptr;
+			};
+
+			inline void insert(SEntry&& entry)
+			{
+				m_container.insert(std::move(entry));
 			}
 
-			virtual IShader::E_CONTENT_TYPE getCodeContentType() const { return IShader::E_CONTENT_TYPE::ECT_UNKNOWN; };
+			// can move to .cpp and have it not inline
+			inline core::smart_refctd_ptr<asset::ICPUShader> find(const SEntry& mainFile, CIncludeFinder* finder)
```

Comment:
`find` should be `const` method

Reply:
btw why the `CIncludeFinder` isn't a const pointer?

Reply:
When I deserialize the Cache I don't immediately load all shaders, they're retrieved from the buffer on demand when someone tries to find its matching entry the first time (they will be loaded on subsequent times or if the shader was compiled at runtime and not fetched from the Cache). Changing `found->value` further down in the code required the method to be non-const I believe. 

I could just change it to load all shaders when the Cache is deserialized otherwise

Reply:
If you want on-demand use `IFileBackedCPUBuffer` with offsets